docs/apis/core/dml/delegated-transactions.md

```diff
@@ -0,0 +1,80 @@
+6. It will be **optional** to catch exceptions when using transactions, but if they are caught, then it is mandatory to mark the transaction for rollback()
+7. Any explicit rollback() call will pass the exception originating it, i.e. rollback($exception) to be re-thrown
+8. API will work in a natural and similar fashion to current recordsets one
+9. We will start **recommending** to use InnoDB for 2.0 in environmental tests. Possibly, at some point in the future that will become a **must**. But **it is not** for now (be warned!)
+
+## The API
+
+1. All the handling must go, exclusively, to a `moodle_database` object, leaving real drivers only implementing (protected) the old begin/commit/rollback_sql() functions
+2. One array of objects of type `moodle_transaction` will be stored / checked from `$DB`
+3. `$DB` will be the responsible to instantiate / accumulate / pair / compare `moodle_transaction`s
+4. Each `moodle_transaction` will be able to set the global mark for rollback. Commit won't change anything
+5. Inner-most commit/rollback will printout one complete stack of `moodle_transaction`s information if we are under `DEBUG_DEVELOPER` and new setting `delegatedtransactionsdebug` is enabled
+6. Normal usage of the moodle_transaction will be:
+
+    ```php
+    $transaction = $DB->start_delegated_transaction();
+    // Perform some $DB stuff
+    $transaction->allow_commit();
+    ```
+
+7. If, for any reason, the developer needs to catch exceptions when using transactions, it will be mandatory to use it in this way:
+
+    ```php
+    try {
+        $transaction = $DB->start_delegated_transaction();
+        // Perform some $DB stuff
+        $transaction->allow_commit();
+    } catch (Exception $e) {
+        //extra cleanup steps
+        $transaction->rollback($e); // rethrows exception
+    }
+    ```
+
+8. In order to be able to keep some parts of code out from top transactions completely, if we know it can lead to problems, we can use:
+
+    ```php
+     $DB->transactions_forbidden(); // Instant check to confirm we aren't using transactions in this point. Will throw exception if transaction is found.
+    ```
+
+## The Flow
+
+![The flow of transactions in diagram format](./_index/TransactionsAndExceptionsFlow.png)
+
+1. Any default exception handler will:
+    1. Catch uncaught transaction_exception exceptions
+    2. Properly perform the DB rollback
+    3. debug/error/log honouring related settings
+    4. inform with as many details as possible (token, place... whatever)
+2. Any "footer" (meaning some place before ending `<html>` output) will:
+    1. Detect "in-transaction" status
+    2. Let execution continue, transaction is automatically rolled back in `$DB->dispose()`
+    3. inform with as many details as possible (token, place... whatever)
+3. `$DB->dispose()` will:
+    1. Detect "in-transaction" status
+    2. log error (not possible to honour settings!)
+    3. Properly perform the full DB rollback
+
+## Related tasks
```

Comment:
Are these tasks still relevant? Are they from when this was a spec?

Ultimately we do not make great use of transactions in Moodle - and I don't know whether we _should_ or not!

1) is obviously done or moot now because txns have been around for so long
2) hopefully we have tests for this!?
3) this probably needs to be made clearer in various places
4)  ??
5) Not really sure if this belongs in this doc

Reply:
I've removed all of these. This has been around long enough that it is either documented or it doesn't make sense.